### PR TITLE
Add flag `isUsingBlockchainContext` to ErgoTree

### PR DESCRIPTION
**Pull request by megatron00999**
Fixes https://github.com/ScorexFoundation/sigmastate-interpreter/issues/928

Add a flag `isUsingBlockchainContext` to ErgoTree similar to the existing `hasDeserialize`.

### CODE REVIEW

**Comment by megatron00999** [comment]
@kushti @aslesarenko Please review and let me know if there is anything else I need to do in this PR. Thank you.

**Comment by aslesarenko** [comment]
@megatron00999, I targeted the next release branch, please resolve the conflicts.

**Review by aslesarenko — CHANGES_REQUESTED** [review]
Please look how _hasDeserialize is initialized
`var _hasDeserialize: Option[Boolean] = givenDeserialize`
and were givenDeserialize comes from.

The key point is [here](https://github.com/ScorexFoundation/sigmastate-interpreter/blob/aae4118fed18f6587413d9a6330e449b05d8d5ad/data/shared/src/main/scala/sigma/serialization/ErgoTreeSerializer.scala#L169) where ErgoTree is created in the serializer with Some(...) value.
This way no additional traverse is necessary after deserialization, which is the main use case during tx validation and is critical for performance.

**Comment by megatron00999** [comment]
@aslesarenko Updated now. Added the detection of blockchain context during parsing in 44844ac15a528fb8afbf132f469468a25237d5dd

File: data/shared/src/main/scala/sigma/ast/values.scala [hunk 1]
```diff
@@ -210,6 +210,30 @@ object Value {
...
 
+  /** Traverses the given expression tree and counts the number of operations
+   * which read the blockchain context. If it is non-zero, returns true. */
+  def isUsingBlockchainContext(exp: SValue): Boolean = {
+    val blockchainContextNode: PartialFunction[Any, Int] = {
+      case Height => 1
+      case LastBlockUtxoRootHash => 1
+      case MinerPubkey => 1
+      case MethodCall(obj, method, _, _) =>
+        (obj, method.objType) match {
```
Comment (aslesarenko) [inline data/shared/src/main/scala/sigma/ast/values.scala:221]: Here, it seems, no need to match on tuple.

File: data/shared/src/main/scala/sigma/serialization/MethodCallSerializer.scala [hunk 2]
```diff
@@ -60,6 +59,16 @@ case class MethodCallSerializer(cons: (Value[SType], SMethod, IndexedSeq[Value[S
...
 
     val specMethod = method.specializeFor(obj.tpe, types)
+
+    var isUsingBlockchainContext = specMethod.objType == SContextMethods && (
+      method.name == SContextMethods.headersMethod.name ||
+        method.name == SContextMethods.preHeaderMethod.name ||
+        method.name == SContextMethods.heightMethod.name ||
+        method.name == SContextMethods.lastBlockUtxoRootHashMethod.name ||
+        method.name == SContextMethods.minerPubKeyMethod.name
+      )
```
Comment (aslesarenko) [inline data/shared/src/main/scala/sigma/serialization/MethodCallSerializer.scala:69]: I think we can slice a bit of performance out of this code. 
The lines like `SContextMethods.headersMethod.name` generate at least 2 method calls. 
I suggest to cache all of them in Array[String] and search in there using tight loop.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: data/shared/src/main/scala/sigma/serialization/MethodCallSerializer.scala (hunk 2 above)
Comment (megatron00999) [reply data/shared/src/main/scala/sigma/serialization/MethodCallSerializer.scala]: Fixed in https://github.com/ScorexFoundation/sigmastate-interpreter/pull/929/commits/22f058625a15767926d53b23ce5812a0e78a7845

Please take a look.

File: data/shared/src/main/scala/sigma/ast/values.scala (hunk 1 above)
Comment (megatron00999) [reply data/shared/src/main/scala/sigma/ast/values.scala]: Updated in https://github.com/ScorexFoundation/sigmastate-interpreter/pull/929/commits/22f058625a15767926d53b23ce5812a0e78a7845.

**Review by aslesarenko — APPROVED** [review]
LGTM, but please fix the tests

**Comment by megatron00999** [comment]
Strangely, `CalcSha256Specification` test failed locally on my system with the following error:

```
[info] CalcSha256Specification:
[info] - CalcSha256: Should pass standard tests. *** FAILED ***
[info]   AssertionError was thrown during property evaluation. (CalcSha256Specification.scala:35)
[info]     Message: assertion failed: Inconsistent stack at List(sigmastate.eval.CProfiler$OpStat@64a05def, sigmastate.eval.CProfiler$OpStat@42d7f5a7, sigmastate.eval.CProfiler$OpStat@643956ec, sigmastate.eval.CProfiler$OpStat@1a3afd0f)
[info]     Occurred at table row 3 (zero based, not counting headings), which had values (
[info]       string = abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu,
[info]       hash = cf5b16a778af8380036ce59e7b0492370b249b11e8f07a51afac45037afee9d1
[info]     )
```

When I reran, the test passed.

**Review by kushti — APPROVED** [review]

**Comment by megatron00999** [comment]
The tests should be fixed now with https://github.com/ScorexFoundation/sigmastate-interpreter/pull/929/commits/0a2390d37d0c4b486ea3d7093107e6c7a8a8b7c6.

**Review by aslesarenko — APPROVED** [review]

**Comment by kushti** [comment]
@megatron00999 please reach me over Discord or email kushti@protonmail.ch and provide ERG address (ordinary or stealth) to get the bounty paid

**Comment by megatron00999** [comment]
@kushti Thanks, sent you a message on Discord with the address.

**Pull request merged**